Some type annotations errors when trying to merge master into my local development branch.

Today, I tried to merge pymatgen master into my local development branch, but met the following type annotations errors:

```shell
(datasci) werner@x13dai-t:~/Public/repo/github.com/materialsproject/pymatgen.git$ git commit -am 'merge master'
[INFO] Checking merge-conflict files only.
ruff.....................................................................Passed
ruff-format..............................................................Passed
check yaml...........................................(no files to check)Skipped
fix end of files.........................................................Passed
trim trailing whitespace.................................................Passed
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

src/pymatgen/symmetry/analyzer.py:1543: error: Incompatible types in assignment (expression has type "Collection[Any]", variable has type "Molecule")  [assignment]
src/pymatgen/io/vasp/sets.py:833: error: Argument 1 to "get_ir_reciprocal_mesh" of "SpacegroupAnalyzer" has incompatible type "tuple[float, float, float] | tuple[int]"; expected "tuple[int, int, int]"  [arg-type]
src/pymatgen/io/vasp/sets.py:867: error: Argument 1 to "get_ir_reciprocal_mesh" of "SpacegroupAnalyzer" has incompatible type "tuple[float, float, float] | tuple[int]"; expected "tuple[int, int, int]"  [arg-type]
src/pymatgen/entries/compatibility.py:558: error: Incompatible types in assignment (expression has type "ComputedEntry | None", variable has type "ComputedEntry")  [assignment]
src/pymatgen/entries/compatibility.py:614: error: Incompatible return value type (got "tuple[ComputedEntry | ComputedStructureEntry, bool]", expected "ComputedEntry | None")  [return-value]
src/pymatgen/entries/compatibility.py:1536: error: List item 0 has incompatible type "Any | list[ComputedEntry | ComputedStructureEntry]"; expected "ComputedEntry"  [list-item]
Found 6 errors in 3 files (checked 2 source files)

codespell................................................................Passed
cython-lint..........................................(no files to check)Skipped
double-quote Cython strings..........................(no files to check)Skipped
blacken-docs.............................................................Passed
markdownlint.........................................(no files to check)Skipped
nbstripout...........................................(no files to check)Skipped
pyright..................................................................Passed
```

The corresponding problematic code lines are shown below: 

```python
werner@x13dai-t:~/Public/repo/github.com/materialsproject/pymatgen.git$ sed -n '1543p' src/pymatgen/symmetry/analyzer.py
        new = eq["sym_mol"]
werner@x13dai-t:~/Public/repo/github.com/materialsproject/pymatgen.git$ sed -n '833p;867p' src/pymatgen/io/vasp/sets.py
                mesh = sga.get_ir_reciprocal_mesh(base_kpoints.kpts[0])
            mesh = sga.get_ir_reciprocal_mesh(zero_weighted_kpoints.kpts[0])
werner@x13dai-t:~/Public/repo/github.com/materialsproject/pymatgen.git$ sed -n '558p;614p;1536p' src/pymatgen/entries/compatibility.py
        entry = self._process_entry_inplace(entry, **kwargs)
        return entry, ignore_entry
            return [entries]
```

I tried some possible modifications, but none of them could pass the pre-commit check. I would be very grateful if you have any suggestions for fixes.

Regards,
Zhao

I could confirm I'm seeing (even more) `mypy` errors on the master branch, I would fix them now.
```
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

src/pymatgen/symmetry/analyzer.py:1543: error: Incompatible types in assignment (expression has type "Collection[Any]", variable has type "Molecule")  [assignment]
src/pymatgen/entries/compatibility.py:558: error: Incompatible types in assignment (expression has type "ComputedEntry | None", variable has type "ComputedEntry")  [assignment]
src/pymatgen/entries/compatibility.py:614: error: Incompatible return value type (got "tuple[ComputedEntry | ComputedStructureEntry, bool]", expected "ComputedEntry | None")  [return-value]
src/pymatgen/entries/compatibility.py:1536: error: List item 0 has incompatible type "Any | list[ComputedEntry | ComputedStructureEntry]"; expected "ComputedEntry"  [list-item]
src/pymatgen/io/packmol.py:216: error: Item "None" of "Molecule | None" has no attribute "to"  [union-attr]
src/pymatgen/analysis/structure_prediction/substitution_probability.py:106: error: Invalid index type "Element | Species" for "dict[Species, float]"; expected type "Species"  [index]
src/pymatgen/io/cp2k/inputs.py:2045: error: Argument "mesh" to "get_ir_reciprocal_mesh" of "SpacegroupAnalyzer" has incompatible type "Sequence[tuple[float, float, float] | tuple[int]]"; expected "tuple[int, int, int]"  [arg-type]
src/pymatgen/io/cp2k/inputs.py:2045: note: Error code "arg-type" not covered by "type: ignore" comment
src/pymatgen/entries/mixing_scheme.py:117: error: Signature of "process_entries" incompatible with supertype "Compatibility"  [override]
src/pymatgen/entries/mixing_scheme.py:117: note:      Superclass:
src/pymatgen/entries/mixing_scheme.py:117: note:          def process_entries(self, entries: ComputedEntry | ComputedStructureEntry | list[ComputedEntry | ComputedStructureEntry], clean: bool = ..., verbose: bool = ..., inplace: bool = ..., n_workers: int = ..., on_error: Literal['ignore', 'warn', 'raise'] = ...) -> list[ComputedEntry | ComputedStructureEntry]
src/pymatgen/entries/mixing_scheme.py:117: note:      Subclass:
src/pymatgen/entries/mixing_scheme.py:117: note:          def process_entries(self, entries: ComputedEntry | ComputedStructureEntry | list[ComputedEntry | ComputedStructureEntry], clean: bool = ..., verbose: bool = ..., inplace: bool = ..., mixing_state_data: Any = ...) -> list[ComputedEntry | ComputedStructureEntry]
src/pymatgen/transformations/site_transformations.py:319: error: Argument 1 to "are_symmetrically_equivalent" of "SpacegroupOperations" has incompatible type "list[PeriodicSite]"; expected "set[PeriodicSite]"  [arg-type]
src/pymatgen/transformations/site_transformations.py:319: error: Argument 2 to "are_symmetrically_equivalent" of "SpacegroupOperations" has incompatible type "list[PeriodicSite]"; expected "set[PeriodicSite]"  [arg-type]
Found 10 errors in 7 files (checked 560 source files)
```

@janosh There're only 6 errors on your side?

1. I'm on a local branch trying to merge the upstream master branch.
2. `entry: env MYPYPATH=src mypy` has already been on the mypy hook.
3. Now, I check out to the pymatgen official master branch, and retest as follows:

```shell
(datasci) werner@x13dai-t:~/Public/repo/github.com/materialsproject/pymatgen.git$ git rev-parse --abbrev-ref `git status | head -1 | awk '{print $NF}'`@{upstream}
upstream/master
(datasci) werner@x13dai-t:~/Public/repo/github.com/materialsproject/pymatgen.git$ git remote -v
origin	https://github.com/hongyi-zhao/pymatgen.git (fetch)
origin	https://github.com/hongyi-zhao/pymatgen.git (push)
upstream	https://github.com/materialsproject/pymatgen.git (fetch)
upstream	https://github.com/materialsproject/pymatgen.git (push)
(datasci) werner@x13dai-t:~/Public/repo/github.com/materialsproject/pymatgen.git$ pre-commit run --all-files
ruff.....................................................................Passed
ruff-format..............................................................Passed
check yaml...............................................................Passed
fix end of files.........................................................Passed
trim trailing whitespace.................................................Passed
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

src/pymatgen/symmetry/analyzer.py:1543: error: Incompatible types in assignment (expression has type "Collection[Any]", variable has type "Molecule")  [assignment]
src/pymatgen/entries/compatibility.py:558: error: Incompatible types in assignment (expression has type "ComputedEntry | None", variable has type "ComputedEntry")  [assignment]
src/pymatgen/entries/compatibility.py:614: error: Incompatible return value type (got "tuple[ComputedEntry | ComputedStructureEntry, bool]", expected "ComputedEntry | None")  [return-value]
src/pymatgen/entries/compatibility.py:1536: error: List item 0 has incompatible type "Any | list[ComputedEntry | ComputedStructureEntry]"; expected "ComputedEntry"  [list-item]
src/pymatgen/io/packmol.py:216: error: Item "None" of "Molecule | None" has no attribute "to"  [union-attr]
src/pymatgen/analysis/structure_prediction/substitution_probability.py:106: error: Invalid index type "Element | Species" for "dict[Species, float]"; expected type "Species"  [index]
src/pymatgen/io/cp2k/inputs.py:2045: error: Argument "mesh" to "get_ir_reciprocal_mesh" of "SpacegroupAnalyzer" has incompatible type "Sequence[tuple[float, float, float] | tuple[int]]"; expected "tuple[int, int, int]"  [arg-type]
src/pymatgen/io/cp2k/inputs.py:2045: note: Error code "arg-type" not covered by "type: ignore" comment
src/pymatgen/entries/mixing_scheme.py:117: error: Signature of "process_entries" incompatible with supertype "Compatibility"  [override]
src/pymatgen/entries/mixing_scheme.py:117: note:      Superclass:
src/pymatgen/entries/mixing_scheme.py:117: note:          def process_entries(self, entries: ComputedEntry | ComputedStructureEntry | list[ComputedEntry | ComputedStructureEntry], clean: bool = ..., verbose: bool = ..., inplace: bool = ..., n_workers: int = ..., on_error: Literal['ignore', 'warn', 'raise'] = ...) -> list[ComputedEntry | ComputedStructureEntry]
src/pymatgen/entries/mixing_scheme.py:117: note:      Subclass:
src/pymatgen/entries/mixing_scheme.py:117: note:          def process_entries(self, entries: ComputedEntry | ComputedStructureEntry | list[ComputedEntry | ComputedStructureEntry], clean: bool = ..., verbose: bool = ..., inplace: bool = ..., mixing_state_data: Any = ...) -> list[ComputedEntry | ComputedStructureEntry]
src/pymatgen/transformations/site_transformations.py:319: error: Argument 1 to "are_symmetrically_equivalent" of "SpacegroupOperations" has incompatible type "list[PeriodicSite]"; expected "set[PeriodicSite]"  [arg-type]
src/pymatgen/transformations/site_transformations.py:319: error: Argument 2 to "are_symmetrically_equivalent" of "SpacegroupOperations" has incompatible type "list[PeriodicSite]"; expected "set[PeriodicSite]"  [arg-type]
Found 10 errors in 7 files (checked 560 source files)

codespell................................................................Passed
cython-lint..............................................................Passed
double-quote Cython strings..............................................Passed
blacken-docs.............................................................Passed
markdownlint.............................................................Passed
nbstripout...............................................................Passed
pyright..................................................................Passed
```

--- a/src/pymatgen/analysis/structure_prediction/substitution_probability.py
+++ b/src/pymatgen/analysis/structure_prediction/substitution_probability.py
@@ -76,7 +76,7 @@ class SubstitutionProbability:
 
         # create Z and px
         self.Z = 0
-        self._px: dict[Species, float] = defaultdict(float)
+        self._px: dict[SpeciesLike, float] = defaultdict(float)
         for s1, s2 in itertools.product(self.species, repeat=2):
             value = math.exp(self.get_lambda(s1, s2))
             self._px[s1] += value / 2

--- a/src/pymatgen/core/periodic_table.py
+++ b/src/pymatgen/core/periodic_table.py
@@ -1609,9 +1609,9 @@ def get_el_sp(obj: SpeciesLike) -> Element | Species | DummySpecies:
 
 @functools.lru_cache
 def get_el_sp(obj: int | SpeciesLike) -> Element | Species | DummySpecies:
-    """Utility method to get an Element, Species or DummySpecies from any input.
+    """Utility function to get an Element, Species or DummySpecies from any input.
 
-    If obj is in itself an Element or a Species, it is returned automatically.
+    If obj is an Element or a Species, it is returned as is.
     If obj is an int or a string representing an integer, the Element with the
     atomic number obj is returned.
     If obj is a string, Species parsing will be attempted (e.g. Mn2+). Failing that
@@ -1626,8 +1626,8 @@ def get_el_sp(obj: int | SpeciesLike) -> Element | Species | DummySpecies:
         ValueError: if obj cannot be converted into an Element or Species.
 
     Returns:
-        Species | Element: with a bias for the maximum number of properties
-            that can be determined.
+        Element | Species | DummySpecies: with a bias for the maximum number
+            of properties that can be determined.
     """
     # If obj is already an Element or Species, return as is
     if isinstance(obj, (Element, Species, DummySpecies)):

--- a/src/pymatgen/core/periodic_table.py
+++ b/src/pymatgen/core/periodic_table.py
@@ -12,7 +12,7 @@ from collections import Counter
 from enum import Enum, unique
 from itertools import combinations, product
 from pathlib import Path
-from typing import TYPE_CHECKING
+from typing import TYPE_CHECKING, overload
 
 import numpy as np
 from monty.dev import deprecated
@@ -1597,11 +1597,21 @@ class DummySpecie(DummySpecies):
     """
 
 
+@overload
+def get_el_sp(obj: int) -> Element:
+    pass
+
+
+@overload
+def get_el_sp(obj: SpeciesLike) -> Element | Species | DummySpecies:
+    pass
+
+
 @functools.lru_cache
 def get_el_sp(obj: int | SpeciesLike) -> Element | Species | DummySpecies:
     """Utility method to get an Element, Species or DummySpecies from any input.
 
-    If obj is in itself an element or a specie, it is returned automatically.
+    If obj is in itself an Element or a Species, it is returned automatically.
     If obj is an int or a string representing an integer, the Element with the
     atomic number obj is returned.
     If obj is a string, Species parsing will be attempted (e.g. Mn2+). Failing that

--- a/src/pymatgen/entries/compatibility.py
+++ b/src/pymatgen/entries/compatibility.py
@@ -1565,7 +1565,6 @@ class MaterialsProjectAqueousCompatibility(Compatibility):
         # Pre-process entries with the given solid compatibility class
         if self.solid_compat:
             entries = self.solid_compat.process_entries(entries, clean=True, inplace=inplace, n_workers=n_workers)
-            return [entries]  # DEBUG: incompatible return type
 
         # when processing single entries, all H2 polymorphs will get assigned the
         # same energy

--- a/src/pymatgen/entries/compatibility.py
+++ b/src/pymatgen/entries/compatibility.py
@@ -35,9 +35,9 @@ from pymatgen.util.joblib import set_python_warnings, tqdm_joblib
 
 if TYPE_CHECKING:
     from collections.abc import Sequence
-    from typing import Literal
+    from typing import Any, Literal
 
-    from pymatgen.util.typing import CompositionLike
+    from pymatgen.util.typing import CompositionLike, PathLike
 
 
 __author__ = "Amanda Wang, Ryan Kingsbury, Shyue Ping Ong, Anubhav Jain, Stephen Dacek, Sai Jayaraman"
@@ -55,7 +55,7 @@ MP_COMPAT_CONFIG = loadfn(f"{MODULE_DIR}/MPCompatibility.yaml")
 # This was compiled by cross-referencing structures in Materials Project from exp_compounds.json.gz
 # used in the fitting of the MP2020 correction scheme, and applying the BVAnalyzer algorithm to
 # determine oxidation state. O and S are not included since these are treated separately.
-MP2020_ANION_OXIDATION_STATE_RANGES = {
+MP2020_ANION_OXIDATION_STATE_RANGES: dict[str, tuple[int, int]] = {
     "Br": (-1, -1),
     "Cl": (-1, -1),
     "F": (-1, -1),
@@ -144,7 +144,12 @@ class PotcarCorrection(Correction):
     'PAW_PBE O 08Apr2002'].
     """
 
-    def __init__(self, input_set: type[VaspInputSet], check_potcar: bool = True, check_hash: bool = False) -> None:
+    def __init__(
+        self,
+        input_set: type[VaspInputSet],
+        check_potcar: bool = True,
+        check_hash: bool = False,
+    ) -> None:
         """
         Args:
             input_set (InputSet): object used to generate the runs (used to check
@@ -170,6 +175,9 @@ class PotcarCorrection(Correction):
         self.input_set = input_set
         self.check_hash = check_hash
         self.check_potcar = check_potcar
+
+    def __str__(self) -> str:
+        return f"{self.input_set.__name__} Potcar Correction"
 
     def get_correction(self, entry: AnyComputedEntry) -> ufloat:
         """
@@ -202,9 +210,6 @@ class PotcarCorrection(Correction):
             raise CompatibilityError(f"Incompatible POTCAR {psp_settings}, expected {expected_psp}")
         return ufloat(0.0, 0.0)
 
-    def __str__(self) -> str:
-        return f"{self.input_set.__name__} Potcar Correction"
-
 
 @cached_class
 class GasCorrection(Correction):
@@ -213,25 +218,28 @@ class GasCorrection(Correction):
     Used by legacy MaterialsProjectCompatibility and MITCompatibility.
     """
 
-    def __init__(self, config_file):
+    def __init__(self, config_file: PathLike) -> None:
         """
         Args:
-            config_file: Path to the selected compatibility.yaml config file.
+            config_file (PathLike): Path to the selected compatibility.yaml config file.
         """
         config = loadfn(config_file)
         self.name = config["Name"]
         self.cpd_energies = config["Advanced"]["CompoundEnergies"]
 
-    def get_correction(self, entry) -> ufloat:
+    def __str__(self) -> str:
+        return f"{self.name} Gas Correction"
+
+    def get_correction(self, entry: ComputedEntry | ComputedStructureEntry) -> ufloat:
         """
         Args:
             entry: A ComputedEntry/ComputedStructureEntry.
 
         Returns:
-            Correction.
+            ufloat: Correction.
         """
         comp = entry.composition
-        # set error to 0 because old MPCompatibility doesn't have errors
+        # Set error to 0 because old MPCompatibility doesn't have errors
         correction = ufloat(0.0, 0.0)
 
         rform = entry.reduced_formula
@@ -239,9 +247,6 @@ class GasCorrection(Correction):
             correction += self.cpd_energies[rform] * comp.num_atoms - entry.uncorrected_energy
 
         return correction
-
-    def __str__(self):
-        return f"{self.name} Gas Correction"
 
 
 @cached_class
@@ -252,11 +257,11 @@ class AnionCorrection(Correction):
     Used by legacy MaterialsProjectCompatibility and MITCompatibility.
     """
 
-    def __init__(self, config_file, correct_peroxide=True):
+    def __init__(self, config_file: PathLike, correct_peroxide: bool = True) -> None:
         """
         Args:
-            config_file: Path to the selected compatibility.yaml config file.
-            correct_peroxide: Specify whether peroxide/superoxide/ozonide
+            config_file (PathLike): Path to the selected compatibility.yaml config file.
+            correct_peroxide (bool): Whether peroxide/superoxide/ozonide
                 corrections are to be applied or not.
         """
         config = loadfn(config_file)
@@ -265,13 +270,16 @@ class AnionCorrection(Correction):
         self.name = config["Name"]
         self.correct_peroxide = correct_peroxide
 
-    def get_correction(self, entry) -> ufloat:
+    def __str__(self) -> str:
+        return f"{self.name} Anion Correction"
+
+    def get_correction(self, entry: ComputedEntry | ComputedStructureEntry) -> ufloat:
         """
         Args:
             entry: A ComputedEntry/ComputedStructureEntry.
 
         Returns:
-            Correction.
+            ufloat: Correction.
         """
         comp = entry.composition
         if len(comp) == 1:  # Skip element entry
@@ -332,9 +340,6 @@ class AnionCorrection(Correction):
 
         return correction
 
-    def __str__(self):
-        return f"{self.name} Anion Correction"
-
 
 @cached_class
 class AqueousCorrection(Correction):
@@ -344,11 +349,11 @@ class AqueousCorrection(Correction):
     Used only by MITAqueousCompatibility.
     """
 
-    def __init__(self, config_file, error_file=None):
+    def __init__(self, config_file: PathLike, error_file: PathLike | None = None) -> None:
         """
         Args:
-            config_file: Path to the selected compatibility.yaml config file.
-            error_file: Path to the selected compatibilityErrors.yaml config file.
+            config_file (PathLike): Path to the selected compatibility.yaml config file.
+            error_file (PathLike): Path to the selected compatibilityErrors.yaml config file.
         """
         config = loadfn(config_file)
         self.cpd_energies = config["AqueousCompoundEnergies"]
@@ -364,13 +369,16 @@ class AqueousCorrection(Correction):
         else:
             self.cpd_errors = defaultdict(float)
 
-    def get_correction(self, entry) -> ufloat:
+    def __str__(self) -> str:
+        return f"{self.name} Aqueous Correction"
+
+    def get_correction(self, entry: ComputedEntry | ComputedStructureEntry) -> ufloat:
         """
         Args:
             entry: A ComputedEntry/ComputedStructureEntry.
 
         Returns:
-            Correction, Uncertainty.
+            ufloat: Correction with uncertainty.
         """
         comp = entry.composition
         rform = comp.reduced_formula
@@ -379,16 +387,14 @@ class AqueousCorrection(Correction):
         correction = ufloat(0.0, 0.0)
 
         if rform in cpd_energies:
-            if rform in ["H2", "H2O"]:
+            if rform in {"H2", "H2O"}:
                 corr = cpd_energies[rform] * comp.num_atoms - entry.uncorrected_energy - entry.correction
-                err = self.cpd_errors[rform] * comp.num_atoms
-
-                correction += ufloat(corr, err)
             else:
                 corr = cpd_energies[rform] * comp.num_atoms
-                err = self.cpd_errors[rform] * comp.num_atoms
 
-                correction += ufloat(corr, err)
+            err = self.cpd_errors[rform] * comp.num_atoms
+            correction += ufloat(corr, err)
+
         if rform != "H2O":
             # if the composition contains water molecules (e.g. FeO.nH2O),
             # correct the gibbs free energy such that the waters are assigned energy=MU_H2O
@@ -421,9 +427,6 @@ class AqueousCorrection(Correction):
 
         return correction
 
-    def __str__(self):
-        return f"{self.name} Aqueous Correction"
-
 
 @cached_class
 class UCorrection(Correction):
@@ -444,21 +447,27 @@ class UCorrection(Correction):
     common_superoxides = ("LiO2", "NaO2", "KO2", "RbO2", "CsO2")
     ozonides = ("LiO3", "NaO3", "KO3", "NaO5")
 
-    def __init__(self, config_file, input_set, compat_type, error_file=None):
+    def __init__(
+        self,
+        config_file: PathLike,
+        input_set,
+        compat_type: Literal["GGA", "Advanced"],
+        error_file: PathLike | None = None,
+    ) -> None:
         """
         Args:
-            config_file: Path to the selected compatibility.yaml config file.
-            input_set: InputSet object (to check for the +U settings)
-            compat_type: Two options, GGA or Advanced. GGA means all GGA+U
-                entries are excluded. Advanced means mixing scheme is
+            config_file (PathLike): Path to the selected compatibility.yaml config file.
+            input_set: InputSet object to check for the +U settings.
+            compat_type ("GGA" | "Advanced"): "GGA" means all GGA+U
+                entries are excluded. "Advanced" means mixing scheme is
                 implemented to make entries compatible with each other,
                 but entries which are supposed to be done in GGA+U will have the
                 equivalent GGA entries excluded. For example, Fe oxides should
                 have a U value under the Advanced scheme. A GGA Fe oxide run
                 will therefore be excluded under the scheme.
-            error_file: Path to the selected compatibilityErrors.yaml config file.
+            error_file (PathLike): Path to the selected compatibilityErrors.yaml config file.
         """
-        if compat_type not in ["GGA", "Advanced"]:
+        if compat_type not in {"GGA", "Advanced"}:
             raise CompatibilityError(f"Invalid {compat_type=}")
 
         config = loadfn(config_file)
@@ -480,13 +489,16 @@ class UCorrection(Correction):
         else:
             self.u_errors = {}
 
-    def get_correction(self, entry) -> ufloat:
+    def __str__(self) -> str:
+        return f"{self.name} {self.compat_type} Correction"
+
+    def get_correction(self, entry: ComputedEntry | ComputedStructureEntry) -> ufloat:
         """
         Args:
             entry: A ComputedEntry/ComputedStructureEntry.
 
         Returns:
-            Correction, Uncertainty.
+            ufloat: Correction with Uncertainty.
         """
         calc_u = entry.parameters.get("hubbards") or defaultdict(int)
         comp = entry.composition
@@ -508,9 +520,6 @@ class UCorrection(Correction):
                 correction += ufloat(u_corr[sym], u_errors[sym]) * comp[el]
 
         return correction
-
-    def __str__(self):
-        return f"{self.name} {self.compat_type} Correction"
 
 
 class Compatibility(MSONable, abc.ABC):
@@ -540,12 +549,17 @@ class Compatibility(MSONable, abc.ABC):
         """
         raise NotImplementedError
 
-    def process_entry(self, entry: ComputedEntry, inplace: bool = True, **kwargs) -> ComputedEntry | None:
-        """Process a single entry with the chosen Corrections. Note
-        that this method will change the data of the original entry.
+    def process_entry(
+        self,
+        entry: ComputedEntry,
+        inplace: bool = True,
+        **kwargs,
+    ) -> ComputedEntry | None:
+        """Process a single entry with the chosen Corrections.
+        Note that this method may change the original entry.
 
         Args:
-            entry: A ComputedEntry object.
+            entry (ComputedEntry): A ComputedEntry object.
             inplace (bool): Whether to adjust the entry in place. Defaults to True.
             **kwargs: Will be passed to process_entries().
 
@@ -555,37 +569,39 @@ class Compatibility(MSONable, abc.ABC):
         if not inplace:
             entry = copy.deepcopy(entry)
 
-        entry = self._process_entry_inplace(entry, **kwargs)
+        _entry: tuple[ComputedEntry, bool] | None = self._process_entry_inplace(entry, **kwargs)
 
-        return entry[0] if entry is not None else None
+        return _entry[0] if _entry is not None else None
 
     def _process_entry_inplace(
         self,
         entry: AnyComputedEntry,
         clean: bool = True,
         on_error: Literal["ignore", "warn", "raise"] = "ignore",
-    ) -> ComputedEntry | None:
-        """Process a single entry with the chosen Corrections. Note
-        that this method will change the data of the original entry.
+    ) -> tuple[ComputedEntry, bool] | None:
+        """Process a single entry with the chosen Corrections.
+        Note that this method will change the original entry.
 
         Args:
-            entry: A ComputedEntry object.
+            entry (AnyComputedEntry): An AnyComputedEntry object.
             clean (bool): Whether to remove any previously-applied energy adjustments.
                 If True, all EnergyAdjustment are removed prior to processing the Entry.
                 Defaults to True.
-            on_error ('ignore' | 'warn' | 'raise'): What to do when get_adjustments(entry)
-                raises CompatibilityError. Defaults to 'ignore'.
+            on_error ("ignore" | "warn" | "raise"): What to do when get_adjustments(entry)
+                raises CompatibilityError. Defaults to "ignore".
 
         Returns:
-            An adjusted entry if entry is compatible, else None.
+            tuple[AnyComputedEntry, ignore_entry (bool)] if entry is compatible, else None.
         """
-        ignore_entry = False
-        # if clean is True, remove all previous adjustments from the entry
+        ignore_entry: bool = False
+        # If clean, remove all previous adjustments from the entry
         if clean:
             entry.energy_adjustments = []
 
-        try:  # get the energy adjustments
-            adjustments = self.get_adjustments(entry)
+        # Get the energy adjustments
+        try:
+            adjustments: list[EnergyAdjustment] = self.get_adjustments(entry)
+
         except CompatibilityError as exc:
             if on_error == "raise":
                 raise
@@ -593,23 +609,27 @@ class Compatibility(MSONable, abc.ABC):
                 warnings.warn(str(exc))
             return None
 
-        for ea in adjustments:
-            # Has this correction already been applied?
-            if (ea.name, ea.cls, ea.value) in [(ea2.name, ea2.cls, ea2.value) for ea2 in entry.energy_adjustments]:
-                # we already applied this exact correction. Do nothing.
+        for e_adj in adjustments:
+            # Check if this correction already been applied
+            if (e_adj.name, e_adj.cls, e_adj.value) in [
+                (ea2.name, ea2.cls, ea2.value) for ea2 in entry.energy_adjustments
+            ]:
+                # We already applied this exact correction. Do nothing.
                 pass
-            elif (ea.name, ea.cls) in [(ea2.name, ea2.cls) for ea2 in entry.energy_adjustments]:
-                # we already applied a correction with the same name
+
+            elif (e_adj.name, e_adj.cls) in [(ea2.name, ea2.cls) for ea2 in entry.energy_adjustments]:
+                # We already applied a correction with the same name
                 # but a different value. Something is wrong.
                 ignore_entry = True
                 warnings.warn(
-                    f"Entry {entry.entry_id} already has an energy adjustment called {ea.name}, but its "
-                    f"value differs from the value of {ea.value:.3f} calculated here. This "
+                    f"Entry {entry.entry_id} already has an energy adjustment called {e_adj.name}, but its "
+                    f"value differs from the value of {e_adj.value:.3f} calculated here. This "
                     "Entry will be discarded."
                 )
+
             else:
                 # Add the correction to the energy_adjustments list
-                entry.energy_adjustments.append(ea)
+                entry.energy_adjustments.append(e_adj)
 
         return entry, ignore_entry
 
@@ -680,8 +700,8 @@ class Compatibility(MSONable, abc.ABC):
         return processed_entry_list
 
     @staticmethod
-    def explain(entry):
-        """Prints an explanation of the energy adjustments applied by the
+    def explain(entry: ComputedEntry) -> None:
+        """Print an explanation of the energy adjustments applied by the
         Compatibility class. Inspired by the "explain" methods in many database
         methodologies.
 
@@ -713,14 +733,18 @@ class CorrectionsList(Compatibility):
     MITCompatibility subclasses instead.
     """
 
-    def __init__(self, corrections: Sequence[Correction], run_types: list[str] | None = None):
+    def __init__(
+        self,
+        corrections: Sequence[Correction],
+        run_types: list[Literal["GGA", "GGA+U", "PBE", "PBE+U"]] | None = None,
+    ) -> None:
         """
         Args:
-            corrections (list[Correction]): Correction objects to apply.
-            run_types: Valid DFT run_types for this correction scheme. Entries with run_type
-                other than those in this list will be excluded from the list returned
-                by process_entries. The default value captures both GGA and GGA+U run types
-                historically used by the Materials Project, for example in.
+            corrections (Sequence[Correction]): Corrections to apply.
+            run_types (list[str]): Valid DFT run types for this correction scheme.
+                Entries with run type other than those in this list will be excluded
+                from the list returned by process_entries. The default value captures
+                both GGA and GGA+U run types historically used by the Materials Project.
         """
         if run_types is None:
             run_types = ["GGA", "GGA+U", "PBE", "PBE+U"]
@@ -745,11 +769,14 @@ class CorrectionsList(Compatibility):
 
         return adjustment_list
 
-    def get_corrections_dict(self, entry: AnyComputedEntry) -> tuple[dict[str, float], dict[str, float]]:
+    def get_corrections_dict(
+        self,
+        entry: AnyComputedEntry,
+    ) -> tuple[dict[str, float], dict[str, float]]:
         """Get the correction values and uncertainties applied to a particular entry.
 
         Args:
-            entry: A ComputedEntry object.
+            entry: An AnyComputedEntry object.
 
         Returns:
             tuple[dict[str, float], dict[str, float]]: Map from correction names to values
@@ -764,7 +791,7 @@ class CorrectionsList(Compatibility):
                 uncertainties[str(c)] = val.std_dev
         return corrections, uncertainties
 
-    def get_explanation_dict(self, entry):
+    def get_explanation_dict(self, entry: ComputedEntry) -> dict[str, Any]:
         """Explain the corrections applied for a given compatibility scheme. Inspired by the
         "explain" methods in many database methodologies.
 
@@ -798,6 +825,8 @@ class CorrectionsList(Compatibility):
                 uncer = np.nan
             else:
                 uncer = uncer_dict.get(str(c), 0)
+
+            assert c.__doc__ is not None
             cd = {
                 "name": str(c),
                 "description": c.__doc__.split("Args")[0].strip(),
@@ -808,8 +837,8 @@ class CorrectionsList(Compatibility):
         dct["corrections"] = corrections
         return dct
 
-    def explain(self, entry):
-        """Prints an explanation of the corrections that are being applied for a
+    def explain(self, entry: ComputedEntry) -> None:  # type: ignore[override]
+        """Print an explanation of the corrections that are being applied for a
         given compatibility scheme. Inspired by the "explain" methods in many
         database methodologies.
 
@@ -839,12 +868,15 @@ class MaterialsProjectCompatibility(CorrectionsList):
     """
 
     def __init__(
-        self, compat_type: str = "Advanced", correct_peroxide: bool = True, check_potcar_hash: bool = False
+        self,
+        compat_type: Literal["GGA", "Advanced"] = "Advanced",
+        correct_peroxide: bool = True,
+        check_potcar_hash: bool = False,
     ) -> None:
         """
         Args:
-            compat_type: Two options, GGA or Advanced. GGA means all GGA+U
-                entries are excluded. Advanced means mixing scheme is
+            compat_type ("GGA" | "Advanced"): "GGA" means all GGA+U
+                entries are excluded. "Advanced" means mixing scheme is
                 implemented to make entries compatible with each other,
                 but entries which are supposed to be done in GGA+U will have the
                 equivalent GGA entries excluded. For example, Fe oxides should
@@ -1522,18 +1554,18 @@ class MaterialsProjectAqueousCompatibility(Compatibility):
             list[AnyComputedEntry]: Adjusted entries. Entries in the original list incompatible with
                 chosen correction scheme are excluded from the returned list.
         """
-        # convert input arg to a list if not already
+        # Convert input arg to a list if not already
         if isinstance(entries, ComputedEntry):
             entries = [entries]
 
-        # if inplace = False, process entries on a copy
+        # If not inplace, process entries on a copy
         if not inplace:
             entries = copy.deepcopy(entries)
 
-        # pre-process entries with the given solid compatibility class
+        # Pre-process entries with the given solid compatibility class
         if self.solid_compat:
             entries = self.solid_compat.process_entries(entries, clean=True, inplace=inplace, n_workers=n_workers)
-            return [entries]
+            return [entries]  # DEBUG: incompatible return type
 
         # when processing single entries, all H2 polymorphs will get assigned the
         # same energy
@@ -1550,25 +1582,26 @@ class MaterialsProjectAqueousCompatibility(Compatibility):
         # do not do this when processing a single entry, as it might lead to unintended
         # results
         if len(entries) > 1:
-            if not self.o2_energy:
-                o2_entries = [e for e in entries if e.reduced_formula == "O2"]
-                if o2_entries:
-                    self.o2_energy = min(e.energy_per_atom for e in o2_entries)
+            if not self.o2_energy and (o2_entries := [e for e in entries if e.reduced_formula == "O2"]):
+                self.o2_energy = min(e.energy_per_atom for e in o2_entries)
 
             if not self.h2o_energy and not self.h2o_adjustments:
-                h2o_entries = [e for e in entries if e.reduced_formula == "H2O"]
-                if h2o_entries:
+                if h2o_entries := [e for e in entries if e.reduced_formula == "H2O"]:
                     h2o_entries = sorted(h2o_entries, key=lambda e: e.energy_per_atom)
                     self.h2o_energy = h2o_entries[0].energy_per_atom
                     self.h2o_adjustments = h2o_entries[0].correction / h2o_entries[0].composition.num_atoms
 
-        h2_entries = [e for e in entries if e.reduced_formula == "H2"]
-        if h2_entries:
+        if h2_entries := [e for e in entries if e.reduced_formula == "H2"]:
             h2_entries = sorted(h2_entries, key=lambda e: e.energy_per_atom)
             self.h2_energy = h2_entries[0].energy_per_atom  # type: ignore[assignment]
 
         return super().process_entries(
-            entries, clean=clean, verbose=verbose, inplace=inplace, n_workers=n_workers, on_error=on_error
+            entries,
+            clean=clean,
+            verbose=verbose,
+            inplace=inplace,
+            n_workers=n_workers,
+            on_error=on_error,
         )
 
 

--- a/src/pymatgen/entries/compatibility.py
+++ b/src/pymatgen/entries/compatibility.py
@@ -1585,7 +1585,7 @@ class MaterialsProjectAqueousCompatibility(Compatibility):
             if not self.o2_energy and (o2_entries := [e for e in entries if e.reduced_formula == "O2"]):
                 self.o2_energy = min(e.energy_per_atom for e in o2_entries)
 
-            if not self.h2o_energy and not self.h2o_adjustments:
+            if not self.h2o_energy and not self.h2o_adjustments:  # noqa: SIM102
                 if h2o_entries := [e for e in entries if e.reduced_formula == "H2O"]:
                     h2o_entries = sorted(h2o_entries, key=lambda e: e.energy_per_atom)
                     self.h2o_energy = h2o_entries[0].energy_per_atom

--- a/src/pymatgen/entries/mixing_scheme.py
+++ b/src/pymatgen/entries/mixing_scheme.py
@@ -114,7 +114,7 @@ class MaterialsProjectDFTMixingScheme(Compatibility):
             if hasattr(compat, "check_potcar"):
                 compat.check_potcar = check_potcar  # type: ignore[union-attr]
 
-    def process_entries(
+    def process_entries(  # type: ignore[override]
         self,
         entries: AnyComputedEntry | list[AnyComputedEntry],
         clean: bool = True,

--- a/src/pymatgen/io/packmol.py
+++ b/src/pymatgen/io/packmol.py
@@ -213,6 +213,7 @@ class PackmolBoxGen(InputGenerator):
                 raise TypeError("Molecule is not provided in supported format.")
 
             fname = f"packmol_{dct['name']}.xyz"
+            assert mol is not None
             mapping[fname] = mol.to(fmt="xyz")
             if " " in str(fname):
                 file_contents.append(f"structure {fname!r}")

--- a/src/pymatgen/symmetry/analyzer.py
+++ b/src/pymatgen/symmetry/analyzer.py
@@ -1503,7 +1503,7 @@ def iterative_symmetrize(
     max_n: int = 10,
     tolerance: float = 0.3,
     epsilon: float = 1e-2,
-) -> dict[Literal["sym_mol", "eq_sets", "sym_ops"], Any]:
+) -> dict[Literal["sym_mol", "eq_sets", "sym_ops"], Molecule | dict]:
     """Get a symmetrized molecule.
 
     The equivalent atoms obtained via `PointGroupAnalyzer.get_equivalent_atoms`

--- a/src/pymatgen/symmetry/analyzer.py
+++ b/src/pymatgen/symmetry/analyzer.py
@@ -1503,47 +1503,41 @@ def iterative_symmetrize(
     max_n: int = 10,
     tolerance: float = 0.3,
     epsilon: float = 1e-2,
-) -> dict:
+) -> dict[Literal["sym_mol", "eq_sets", "sym_ops"], Any]:
     """Get a symmetrized molecule.
 
-    The equivalent atoms obtained via
-    :meth:`~pymatgen.symmetry.analyzer.PointGroupAnalyzer.get_equivalent_atoms`
+    The equivalent atoms obtained via `PointGroupAnalyzer.get_equivalent_atoms`
     are rotated, mirrored... unto one position.
-    Then the average position is calculated.
-    The average position is rotated, mirrored... back with the inverse
-    of the previous symmetry operations, which gives the
-    symmetrized molecule
+    Then the average position is calculated, which is rotated, mirrored...
+    back with the inverse of the previous symmetry operations, giving the
+    symmetrized molecule.
 
     Args:
         mol (Molecule): A pymatgen Molecule instance.
         max_n (int): Maximum number of iterations.
-        tolerance (float): Tolerance for detecting symmetry.
-            Gets passed as Argument into
-            ~pymatgen.analyzer.symmetry.PointGroupAnalyzer.
+        tolerance (float): Tolerance for detecting symmetry with PointGroupAnalyzer.
         epsilon (float): If the element-wise absolute difference of two
             subsequently symmetrized structures is smaller epsilon,
             the iteration stops before max_n is reached.
 
     Returns:
-        dict: with three possible keys:
-            sym_mol: A symmetrized molecule instance.
+        dict with three keys:
+            sym_mol: A symmetrized Molecule instance.
             eq_sets: A dictionary of indices mapping to sets of indices, each key maps to indices
                 of all equivalent atoms. The keys are guaranteed to be not equivalent.
-            sym_ops: Twofold nested dictionary. operations[i][j] gives the symmetry operation
+            sym_ops: Two-fold nested dictionary. operations[i][j] gives the symmetry operation
                 that maps atom i unto j.
     """
-    new = mol
-    n = 0
-    finished = False
-    eq = {"sym_mol": new, "eq_sets": {}, "sym_ops": {}}
-    while not finished and n <= max_n:
-        previous = new
-        PA = PointGroupAnalyzer(previous, tolerance=tolerance)
-        eq = PA.symmetrize_molecule()
-        new = eq["sym_mol"]
-        finished = np.allclose(new.cart_coords, previous.cart_coords, atol=epsilon)
-        n += 1
-    return eq
+    new_mol: Molecule = mol
+    sym_mol: dict = {"sym_mol": new_mol, "eq_sets": {}, "sym_ops": {}}
+    for _ in range(max_n):
+        prev_mol: Molecule = new_mol
+        sym_mol = PointGroupAnalyzer(prev_mol, tolerance=tolerance).symmetrize_molecule()
+        new_mol = sym_mol["sym_mol"]
+
+        if np.allclose(new_mol.cart_coords, prev_mol.cart_coords, atol=epsilon):
+            break
+    return sym_mol
 
 
 def cluster_sites(

--- a/src/pymatgen/transformations/site_transformations.py
+++ b/src/pymatgen/transformations/site_transformations.py
@@ -316,7 +316,7 @@ class PartialRemoveSitesTransformation(AbstractTransformation):
             for ii, t_sites in enumerate(tested_sites):
                 t_energy = all_structures[ii]["energy"]
                 if abs((energy - t_energy) / len(s_new)) < 1e-5 and sg.are_symmetrically_equivalent(
-                    sites_to_remove, t_sites, symm_prec=symprec
+                    set(sites_to_remove), set(t_sites), symm_prec=symprec
                 ):
                     already_tested = True
 

--- a/src/pymatgen/util/io_utils.py
+++ b/src/pymatgen/util/io_utils.py
@@ -4,8 +4,12 @@ from __future__ import annotations
 
 import os
 import re
+from typing import TYPE_CHECKING
 
 from monty.io import zopen
+
+if TYPE_CHECKING:
+    from collections.abc import Generator
 
 __author__ = "Shyue Ping Ong, Rickard Armiento, Anubhav Jain, G Matteo, Ioannis Petousis"
 __copyright__ = "Copyright 2011, The Materials Project"
@@ -16,7 +20,7 @@ __status__ = "Production"
 __date__ = "Sep 23, 2011"
 
 
-def clean_lines(string_list, remove_empty_lines=True) -> list[str]:
+def clean_lines(string_list, remove_empty_lines=True) -> Generator[str, None, None]:
     """Strips whitespace, carriage returns and empty lines from a list of strings.
 
     Args:

